backend/danswer/datastores/interfaces.py

```diff
@@ -1,22 +1,42 @@
 import abc
+from typing import Generic
+from typing import TypeVar
 
+from danswer.chunking.models import BaseChunk
 from danswer.chunking.models import EmbeddedIndexChunk
+from danswer.chunking.models import IndexChunk
 from danswer.chunking.models import InferenceChunk
 
-DatastoreFilter = dict[str, str | list[str] | None]
 
+T = TypeVar("T", bound=BaseChunk)
+IndexFilter = dict[str, str | list[str] | None]
 
-class Datastore:
+
+class DocumentIndex(Generic[T], abc.ABC):
     @abc.abstractmethod
-    def index(self, chunks: list[EmbeddedIndexChunk], user_id: int | None) -> bool:
+    def index(self, chunks: list[T], user_id: int | None) -> bool:
         raise NotImplementedError
 
+
+class VectorIndex(DocumentIndex[EmbeddedIndexChunk], abc.ABC):
     @abc.abstractmethod
     def semantic_retrieval(
```

Comment:
nit: `semantic_retrieval` -> `semantic_search` OR change `keyword_search` -> `keyword_retrieval` for consistency